arroyo/processing/strategies/batching.py

```diff
@@ -262,3 +265,265 @@ def create_with_partitions(
+    downstream steps if they are thrown.
+    """
+
+    def __init__(
+        self,
+        max_batch_time_sec: float,
+        max_batch_size: int,
+        next_step: ProcessingStrategy[MessageBatch[TPayload]],
+    ) -> None:
+        self.__max_batch_time_sec = max_batch_time_sec
+        self.__max_batch_size = max_batch_size
+        self.__next_step = next_step
+        self.__batch_builder: Optional[BatchBuilder[TPayload]] = None
+        self.__closed = False
+
+    def __flush(self) -> None:
+        assert self.__batch_builder is not None
+        batch = self.__batch_builder.build_if_ready()
+        if batch is None:
+            return
+
+        last_msg = batch.last()
+        if last_msg is None:
+            # TODO: PR #134 will fix this problem and make it unnecessary
+            # to hack partition id = 0 and offset = 0 for empty batches.
+            batch_msg = Message(
+                partition=Partition(Topic(""), 0),
+                offset=0,
+                timestamp=datetime.now(),
+                payload=batch,
+            )
+        else:
+            batch_msg = Message(
+                partition=last_msg.partition,
+                offset=last_msg.offset,
+                timestamp=last_msg.timestamp,
+                payload=batch,
+            )
+        self.__next_step.submit(batch_msg)
+        self.__batch_builder = None
+
+    def submit(self, message: Message[TPayload]) -> None:
+        """
+        Accumulates messages in the current batch.
+        A new batch is created at the first message received.
+
+        This method tries to flush before adding the message
+        to the current batch. This is so that, if we receive
+        `MessageRejected` exception from the following step,
+        we can propagate the exception without processing the
+        new message. This allows the previous step to try again
+        without introducing duplications.
+        """
+        assert not self.__closed
+
+        if self.__batch_builder is not None:
+            self.__flush()
+
+        if self.__batch_builder is None:
+            self.__batch_builder = BatchBuilder(
+                max_batch_time=self.__max_batch_time_sec,
+                max_batch_size=self.__max_batch_size,
+            )
+
+        self.__batch_builder.append(message)
+
+    def poll(self) -> None:
+        assert not self.__closed
+
+        if self.__batch_builder is not None:
+            try:
+                self.__flush()
+            except MessageRejected:
+                pass
+
+        self.__next_step.poll()
+
+    def close(self) -> None:
+        self.__closed = True
+
+    def terminate(self) -> None:
+        self.__closed = True
+        self.__batch_builder = None
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        """
+        Terminates the strategy by joining the following step.
+        This method throws away the current batch.
```

Comment:
In the previous implementation I was flushing the batch here. 
Though I noticed some inconsistency in how we manage the `join` in the code base. Some steps like the tickbuffer or the parallel consumer abandon the work they were doing, while the collector flushes it. 

I think we need to get decide how to do this and stick to it.
What do you think? Ideally I would prefer flushing on join but then I get stuck in how to deal with MessageRejected. The join would have to loop till the message gets accepted potentially hitting the deadline and preventing the join of the following step from happening.